New event handling system.

whichonespink44 opened the pull request:
So I've been looking into RTG's event management issues, and I think this might be a decent approach to use, so I'm uploading this for review.

Essentially, instead of having all of the event handlers as methods in one class (EventManagerRTG), we need to move them out into their own classes (or as inner classes within EventManagerRTG). Doing this allows us to use the World.Load and World.Unload events to register/unregister the handlers, ensuring that RTG's event handlers are only registered in the RTG world type. The only event handlers that are ALWAYS registered are the World.Load and World.Unload handlers because they're the ones doing the registering and unregistering.

So the process looks something like this:
- preInit -> Register all event handlers.
- onWorldLoad -> Register all event handlers (if they're already registered, they don't get added twice)
- onWorldUnload -> Unregister all event handlers (except onWorldLoad and onWorldUnload)

This accounts for all the different scenarios where RTG's event handlers should/shouldn't be registered. The only place this approach might fall down is when dealing with multi-world servers, but I'm not exactly sure how that stuff works, so maybe there's a way around it, maybe there isn't.

In any case, there's still a bit of work to do with this PR, ~~for example, decorations aren't working properly (but I think I know why - just haven't gotten around to fixing it yet)~~, but I thought I'd put it up for review anyway to see what everyone thinks.

If this is the approach we end up using, it's going to require extensive snapshot testing, and we should probably also bump the version number to 1.1.0.

whichonespink44 committed:
Added new event handling system.

whichonespink44 committed:
Event handlers are now inner classes within EventManagerRTG.

Doing this has also resolved the decoration issue, but further testing
still required.

srs-bsns committed:
Backported @topisani's Accessor changes in 1.9.4

srs-bsns committed:
Updated build.gradle with dubugger settings

whichonespink44 committed:
Merge pull request #821 from srs-bsns/1.7.10/event-manager

Dev environment stuff.

srs-bsns committed:
More Event Manager refinement

whichonespink44 committed:
Merge pull request #826 from srs-bsns/1.7.10/event-manager

More Event Manager refinement.

srs-bsns commented:
FYI: This will revert a2fb7b82fface7096c1e652a6279c7df81917ffa which fixed #392, however, the issue no longer seems relevant.

srs-bsns committed:
Fix for AdminCommandsToolbox causing event handlers to unregister on running server

whichonespink44 committed:
Merge pull request #827 from srs-bsns/1.7.10/event-manager

Fix for AdminCommandsToolbox causing event handlers to unregister on running server.

whichonespink44 committed:
Removed @SuppressWarnings

whichonespink44 committed:
Removed @SuppressWarnings

whichonespink44 committed:
Added 'receiveCanceled' to event handlers.

Logging cancelled events ahead of snapshot release to test for other
mods interfering with RTG event handlers.

Zeno410 commented:
Generally having the event handlers as separate objects is a good thing. There is a problem with registering on world load, though, at least in 1.7.10: some generation is done in the instantiation of the world object, before it gets loaded (I have had many tears over this insane design decision). Also, biome gen init occurs earlier, so if we don't stick to our stringent "don't futz with the biome generator" policy that might cause problems too.

srs-bsns commented:
> There is a problem with registering on world load, though, at least in 1.7.10: some generation is done in the instantiation of the world object

Yeah, that crap had me pulling my hair out for a while. I couldn't figure out why a certain event wasn't firing when the event class was clearly being registered. It wasn't until later that I realised I had to register the event at an earlier point in order for it to actually receive the events it needed to do it's thing. The event system isn't very intuitive at all.

whichonespink44 committed:
Merge remote-tracking branch 'origin/1.7.10/event-manager' into 1.7.10/event-manager

# Conflicts: RESOLVED
#	src/main/java/rtg/event/EventManagerRTG.java

whichonespink44 committed:
Removed 'receiveCanceled' & minor refactoring.

None of the events that RTG subscribes to are cancellable so
'receiveCanceled' was pointless. Also, the minor refactoring was mostly
so that I could get my head into the new system as the new handler
names, whilst totally fine, were making it difficult for me to see what
was going on, as I was so used to the other naming convention. (Sorry
@srs-bsns !!!)
